File: src/main/scala/xenit/care4alf/module/WorkflowInstances.scala
package xenit.care4alf.module

import org.springframework.beans.factory.annotation.Autowired
import xenit.care4alf.web.{JsonHelper, Json}
import com.github.dynamicextensionsalfresco.webscripts.annotations._

import scala.collection.JavaConversions._
import org.springframework.stereotype.Component
import org.alfresco.service.cmr.workflow.{WorkflowTaskQuery, WorkflowInstance, WorkflowService}
import com.typesafe.scalalogging.slf4j.Logging
import org.alfresco.model.ContentModel
import xenit.care4alf.alfresco.{HasNamespaceService, HasNodeService}
import org.alfresco.repo.admin.SysAdminParams
import org.alfresco.util.UrlUtil
import org.springframework.util.StringUtils

import org.alfresco.repo.security.authentication.AuthenticationUtil.runAsSystem
import xenit.care4alf.alfresco.Implicits._

/**
 * @author Laurent Van der Linden
 */
@Component
@WebScript(baseUri = "/xenit/care4alf/workflow/instances", families = Array("care4alf"), description = "Workflow instances")
@Authentication(AuthenticationType.ADMIN)
class WorkflowInstances extends Json with Logging with HasNodeService with HasNamespaceService with RestErrorHandling {
    @Autowired var workflowService: WorkflowService = null
    @Autowired var sysadminParams: SysAdminParams = null

    @Uri(value = Array("/active"), defaultFormat = "json")
    def activeInstances(@Attribute jsonHelper: JsonHelper, @RequestParam(defaultValue = "includeTasks") includeTasks: Boolean) {
        val json = jsonHelper.json
        json.array()
        for (instance <- workflowService.getActiveWorkflows.filter(_.getDefinition != null)) {
            instanceToJson(instance, jsonHelper, includeTasks)
        }
        json.endArray()
    }

    @Uri(value = Array("/find/task/{taskid}"), defaultFormat = "json")
    def instanceByTask(@Attribute jsonHelper: JsonHelper, @UriVariable taskid: String) {
        val json = jsonHelper.json
        val workflowTask = runAsSystem {
            workflowService.getTaskById(taskid)
        }
        json.array()
        instanceToJson(workflowTask.getPath.getInstance(), jsonHelper, includeTasks = true)
        json.endArray()
    }

    @Uri(value = Array("/find/instance/{instanceid}"), defaultFormat = "json")
    def instanceById(@Attribute jsonHelper: JsonHelper, @UriVariable instanceid: String) {
        val instance = workflowService.getWorkflowById(instanceid)
        if (instance != null) {
            val json = jsonHelper.json
            json.array()
            instanceToJson(instance, jsonHelper, includeTasks = true)
            json.endArray()
        } else {
            throw new IllegalArgumentException(s"No workflow instance $instanceid found.")
        }
    }


    def instanceToJson(instance: WorkflowInstance, jsonHelper: JsonHelper, includeTasks: Boolean) {
        val json = jsonHelper.json
        val initiator = instance.getInitiator
        val initiatorUsername = if (nodeService.exists(initiator)) initiator(ContentModel.PROP_USERNAME).asInstanceOf[String] else "-"
        val description = if (StringUtils.hasText(instance.getDescription)) instance.getDescription else instance.getDefinition.getDescription
        json.`object`()
            .key("description").value(description)
            .key("id").value(instance.getId)
            .key("initiator").value(initiatorUsername)
            .key("start").value(instance.getStartDate.getTime)
            .key("files").array()
        for (assoc <- nodeService.getChildAssocs(instance.getWorkflowPackage)) {
            json.`object`()
                .key("path").value(assoc.getChildRef.path)
                .key("url").value(UrlUtil.getShareUrl(sysadminParams) + "/page/document-details?nodeRef=" + assoc.getChildRef)
            .endObject()
        }
        json.endArray()
        if (includeTasks) {
            json.key("tasks")
            tasksForInstance(jsonHelper, instance.getId)
        }
        json.endObject()
    }

    @Uri(value = Array("/{workflowid}/tasks"), defaultFormat = "json")
    def tasksForInstance(@Attribute jsonHelper: JsonHelper, @UriVariable workflowid: String) {
        val json = jsonHelper.json
        val tasks = runAsSystem {
            getTasksForInstance(workflowService.getWorkflowById(workflowid))
        }
        json.array()
        for (task <- tasks) {
            json.`object`()
                .key("id").value(task.getId)
                .key("description").value(task.getDescription)
                .key("properties").`object`()
            for (prop <- task.getProperties) {
                json.key(prop._1.toString).value(prop._2)
            }
            json.endObject()
                .endObject()
        }
        json.endArray()
    }

    def getTasksForInstance(instance: WorkflowInstance) = {
        val query = new WorkflowTaskQuery
        query.setProcessId(instance.getId)
        query.setTaskState(null)
        query.setActive(null)
        workflowService.queryTasks(query)
    }

    @Uri(value = Array("/{id}/cancel"), method = HttpMethod.DELETE)
    def cancelWorkflow(@UriVariable("id") id: String) {
        workflowService.cancelWorkflow(id)
    }

    @Uri(value = Array("/{id}/delete"), method = HttpMethod.DELETE)
    def deleteWorkflow(@UriVariable("id") id: String) {
        workflowService.deleteWorkflow(id)
    }

    @Uri(value = Array("/active"), method = HttpMethod.DELETE)
    def deleteAllActive() {
        for (wf <- workflowService.getActiveWorkflows) {
            try {
                workflowService.deleteWorkflow(wf.getId)
                logger.debug(s"Deleted workflow instance ${wf.getId}.")
            }
            catch {
                case ex: Exception => {
                    logger.info(s"Failed to delete workflow instance ${wf.getId}: ${ex.getMessage}.")
                }
            }
        }
    }
}

File: src/main/kotlin/eu/xenit/care4alf/module/WorkflowInstances.kt
package eu.xenit.care4alf.module

import com.github.dynamicextensionsalfresco.webscripts.annotations.*
import eu.xenit.care4alf.JsonRoot
import eu.xenit.care4alf.json
import org.alfresco.model.ContentModel
import org.alfresco.repo.admin.SysAdminParams
import org.alfresco.repo.security.authentication.AuthenticationUtil
import org.alfresco.service.cmr.repository.NodeService
import org.alfresco.service.cmr.security.PermissionService
import org.alfresco.service.cmr.workflow.WorkflowInstance
import org.alfresco.service.cmr.workflow.WorkflowService
import org.alfresco.service.cmr.workflow.WorkflowTask
import org.alfresco.service.cmr.workflow.WorkflowTaskQuery
import org.alfresco.service.namespace.NamespacePrefixResolver
import org.alfresco.service.namespace.QName
import org.alfresco.util.ISO8601DateFormat
import org.alfresco.util.UrlUtil
import org.json.JSONObject
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import org.springframework.util.StringUtils
import java.io.Serializable
import java.util.*

/**
 * @author Laurent Van der Linden
 */
@Component
@WebScript(baseUri = "/xenit/care4alf/workflow/instances", families = arrayOf("care4alf"), description = "Workflow instances")
@Authentication(AuthenticationType.ADMIN)
public class WorkflowInstances @Autowired constructor(
        private val workflowService: WorkflowService,
        private val permissionService: PermissionService,
        private val sysadminParams: SysAdminParams,
        private val nodeService: NodeService,
        private val nameSpacePrefixResolver: NamespacePrefixResolver
    ) {

    private val logger = LoggerFactory.getLogger(javaClass)

    @Uri(value = "/active", defaultFormat = "json")
    fun activeInstances() = json {
        iterable(workflowService.getActiveWorkflows(), instanceToJson(false))
    }

    @Uri(value = "/find/task/{taskid}", defaultFormat = "json")
    fun instanceByTask(@UriVariable taskid: String) = json {
        val workflowTask = AuthenticationUtil.runAsSystem {
            workflowService.getTaskById(taskid)
        }
        iterable(listOf(workflowTask.getPath().getInstance()), instanceToJson(includeTasks = true))
    }

    @Uri(value = "/find/instance/{instanceid}", defaultFormat = "json")
    fun instanceById(@UriVariable instanceid: String) = json {
        iterable(listOf(workflowService.getWorkflowById(instanceid)), instanceToJson(includeTasks = true))
    }

    fun instanceToJson(includeTasks: Boolean): JsonRoot.(WorkflowInstance) -> Unit {
        return { instance ->
            val initiator = instance.getInitiator()
            val initiatorUsername = if (nodeService.exists(initiator)) {
                nodeService.getProperty(initiator, ContentModel.PROP_USERNAME) as String
            } else {
                "-"
            }
            val description = if (StringUtils.hasText(instance.getDescription())) {
                instance.getDescription()
            } else {
                instance.getDefinition().getDescription()
            }

            obj {
                entry("id", instance.getId())
                entry("description", description)
                entry("initiator", initiatorUsername)
                entry("start", instance.getStartDate()?.getTime())
                key("definition") {
                    obj {
                        entry("id", instance.getDefinition().getId())
                        entry("name", instance.getDefinition().getName())
                    }
                }
                key("files") {
                    val children = nodeService.getChildAssocs(instance.getWorkflowPackage()).map({it.getChildRef()})
                    iterable(children) { file ->
                        obj {
                            entry("path", nodeService.getPath(file).toDisplayPath(nodeService, permissionService))
                            entry("url", UrlUtil.getShareUrl(sysadminParams) + "/page/document-details?nodeRef=" + file)
                        }
                    }
                }
                if (includeTasks) {
                    val tasks = AuthenticationUtil.runAsSystem {
                        getTasksForInstance(instance)
                    }
                    key("tasks") {
                        iterable(tasks, taskToJson())
                    }
                }
            }
        }
    }

    fun taskToJson(): JsonRoot.(WorkflowTask) -> Unit {
        return {
            task -> obj {
                entry("id", task.getId())
                entry("name", task.getName())
                entry("title", task.getTitle())
                entry("state", task.getState())
                entry("description", task.getDescription())
                key("properties") {
                    obj {
                        for (property in task.getProperties()) {
                            entry(property.key.toString(), property.value)
                        }
                    }
                }
            }
        }

    }

    @Uri(value = "/{workflowid}/tasks", defaultFormat = "json")
    fun tasksForInstance(@UriVariable workflowid: String) = json {
        val tasks = AuthenticationUtil.runAsSystem {
            getTasksForInstance(workflowService.getWorkflowById(workflowid))
        }
        iterable(tasks, taskToJson())
    }

    fun getTasksForInstance(instance: WorkflowInstance): List<WorkflowTask>? {
        val query = WorkflowTaskQuery()
        query.setProcessId(instance.getId())
        query.setTaskState(null)
        query.setActive(null)
        return workflowService.queryTasks(query)
    }

    @Uri(value = "/{id}/cancel", method = HttpMethod.DELETE)
    fun cancelWorkflow(@UriVariable("id") id: String) {
        workflowService.cancelWorkflow(id)
    }

    @Uri(value = "/{id}/delete", method = HttpMethod.DELETE)
    fun deleteWorkflow(@UriVariable("id") id: String) {
        workflowService.deleteWorkflow(id)
    }

    @Uri(value = "/active", method = HttpMethod.DELETE)
    fun deleteAllActive() {
        for (wf in workflowService.getActiveWorkflows()) {
            try {
                workflowService.deleteWorkflow(wf.getId())
                logger.debug("Deleted workflow instance ${wf.getId()}.")
            }
            catch(ex: Exception) {
                logger.info("Failed to delete workflow instance ${wf.getId()}: ${ex.message}.")
            }
        }
    }

    @Uri(value = "/tasks/{id}/release", method = HttpMethod.POST, defaultFormat = "json")
    fun releaseTask(@UriVariable("id") id: String) = json {
        logger.error("Id is {}", id);
        val props: MutableMap<QName, Serializable?> = hashMapOf(ContentModel.PROP_OWNER to null);
        val task = workflowService.updateTask(id, props, null, null);
        obj {
            entry("id", task.getId())
            entry("description", task.getDescription())
            key("properties") {
                obj {
                    for (property in task.getProperties()) {
                        entry(property.key.toString(), property.value)
                    }
                }
            }
        }
    }

    @Uri(value = "/tasks/{id}/setProperty", method = HttpMethod.POST, defaultFormat = "json")
    fun setTaskProperty(@UriVariable("id") id: String, payload: JSONObject) = json {
        logger.info("Setting property on {}", id);
        val type = payload.getString("type");
        val qname : QName = QName.createQName(payload.getString("qname"), nameSpacePrefixResolver);
        var props: MutableMap<QName, Serializable?> = hashMapOf();
        when (type) {
            "Date" -> {
                val date: Date = ISO8601DateFormat.parse(payload.getString("value"));
                props = hashMapOf(qname to date);
            }
            "Integer" -> {
                val integer: Int = Integer.parseInt(payload.getString("value"));
                props = hashMapOf(qname to integer);
            }
            "String" -> props = hashMapOf(qname to payload.getString("value"));
        }
        if (props.isEmpty()) {
            throw IllegalArgumentException("Type $type not recognized, should be String, Integer or Date");
        }
        val task = workflowService.updateTask(id, props, null, null);
        obj {
            entry("id", task.getId())
            entry("description", task.getDescription())
            key("properties") {
                obj {
                    for (property in task.getProperties()) {
                        entry(property.key.toString(), property.value)
                    }
                }
            }
        }
    }
}